akka-stream-tests/src/test/scala/akka/stream/scaladsl/SealedAttributesSpec.scala

```diff
@@ -0,0 +1,253 @@
+      override def initialAttributes = initialAttr
+
+      override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+        val logic = new GraphStageLogic(shape) with InHandler {
+
+          override def onPush(): Unit = ()
+
+          setHandler(shape.in, this)
+        }
+
+        logic → get(inheritedAttributes)
+      }
+    })
+
+  def graphStageSinkShape[T, A](get: Attributes ⇒ Option[A], initialAttr: Attributes = Attributes.none): GraphStageWithMaterializedValue[SinkShape[Any], Option[A]] =
+    new GraphStageWithMaterializedValue[SinkShape[Any], Option[A]] {
+      override val shape: SinkShape[Any] = SinkShape.of(Inlet[Any]("in"))
+      override def initialAttributes = initialAttr
+
+      override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+        val logic = new GraphStageLogic(shape) with InHandler {
+
+          println(s" >>> inheritedAttributes = ${inheritedAttributes}")
+
+          val dispatchers = inheritedAttributes.filtered[ActorAttributes.Dispatcher].map(_.dispatcher)
+          println(s" >>> dispatchers = ${dispatchers}")
+
+          //          val names = inheritedAttributes.filtered[Name].map(_.n)
+          //          println(s" >>> names = ${names.mkString(" nested in: ")}")
+          override def onPush(): Unit = ()
+
+          setHandler(shape.in, this)
+        }
+
+        logic → get(inheritedAttributes)
+      }
+    }
+
+}
+
+class SealedAttributesSpec extends StreamSpec(
+  s"""
+    akka.actor.default-mailbox.mailbox-type = "akka.dispatch.UnboundedMailbox"
+
+    dispatcher-outer = $${akka.actor.default-dispatcher}
+    dispatcher-default-in-stage-initial = $${akka.actor.default-dispatcher}
+    right-on-sink = $${akka.actor.default-dispatcher}
+    completely-outside = $${akka.actor.default-dispatcher}
+    outer = $${akka.actor.default-dispatcher}
+  """.stripMargin) with ScalaFutures {
+  import SealedAttributesSpec._
+
+  val settings = ActorMaterializerSettings(system)
+    .withInputBuffer(initialSize = 2, maxSize = 16)
+
+  implicit val materializer = ActorMaterializer(settings)
+
+  def attributesOf[T](s: RunnableGraph[Attributes]): Attributes = {
+    s.run()
+  }
+
+  def attributesOf[T](s: RunnableGraph[Option[T]]): Option[T] = {
+    s.run()
+  }
+
+  "attributes.get" must {
+    "return None if not present" in {
+      Attributes().get[Name] should ===(None)
+    }
+
+    "return it when present, trivial case " in {
+      val attrs = Attributes(Name("hello"))
+      attrs.get[Name] should ===(Some(Name("hello")))
+    }
+
+    "return override, when has 2 values, no seals" in {
+      val attrs = Attributes(Name("hello") :: Name("default-name") :: Nil)
+      attrs.get[Name] should ===(Some(Name("hello")))
+    }
+
+    "return sealed, when has 2 values, default is sealed" in {
+      val attrs = Attributes(Name("hello") :: Name("default-name").seal("name") :: Nil)
```

Comment:
I really don't like the shared-string-identifier idea. Could it be for the attribute type instead?

Reply:
Yeah, absolutely -- was thinking that. It's either names or types.
If we do it as a type it could be a matcher I guess... it can return yes/no if it should overwrite or not hm... This loops back to @jrudolph's idea about the attributes having some say in how they are to be handled

Reply:
I got worried that it gets too complex/powerful.
The boring names you can understand what's happening by just comparing printouts and `==` ing names and the `*` hm